src/Servers/Kestrel/shared/CompletionPipeReader.cs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO.Pipelines;
+
+#nullable enable
+
+namespace Microsoft.AspNetCore.Connections
+{
+    internal sealed class CompletionPipeReader : PipeReader
+    {
+        private readonly PipeReader _inner;
+
+        public bool IsComplete { get; private set; }
+        public Exception? CompleteException { get; private set; }
+        public bool IsCompleteSuccessfully => IsComplete && CompleteException == null;
```

Comment:
Nit: `HasCompletedSuccessfully`?

Reply:
Changed to IsCompletedSuccessfully. I indented to make it consistent with Task.IsCompletedSuccessfully.